### PR TITLE
add event API

### PR DESCRIPTION
**Pull request by williamboman**
Opening a WIP draft PR just to get a discussion going.

I think it'd be really neat being able to hook into the different actions nvim-tree does. For example - when renaming a node I'd like to inform my LSP client that a rename just happened:

```lua
local events = require'nvim-tree.events'
local typescript_ls = require'lsp.clients.typescript'

events.register_callback('renamed', function (old_fname, new_fname)
    typescript_ls.rename_file(old_fname, new_fname)
end)
```

I went with quite a vanilla event system where any amount of callbacks can be registered for any given event name.

### CODE REVIEW

**Comment by kyazdani42** [comment]
Hi i think that would be a great feature to add indeed, thanks for implementing the first draw, you could go for it i'd accept a fully made PR :)

**Comment by williamboman** [comment]
Question: When deleting a directory, would you as a consumer expect to receive a) a single "folder deleted" event, or b) an event for each file and dir deleted within the deleted folder?

**Comment by kyazdani42** [comment]
I guess i'd prefer having the main folder and also all deleted files. As a consumer, i wouldn't expect reproducing what the plugin already did :)

**Comment by williamboman** [comment]
I went with a simpler approach where events are structured around the user action itself. This might be a bit too opaque for certain use cases, but I figured it's a good start.

With the current implementation, you can't really figure out which files were deleted or "renamed" if you are deleting or renaming a folder, you just get the FolderRenamed/FolderRemoved events. At least not out of the box.

File: doc/nvim-tree-lua.txt [hunk 1]
```diff
@@ -1,6 +1,6 @@
 *nvim-tree.lua* A file explorer written in lua
 
-Minimum version of neovim: 0.4.0
+Minimum version of neovim: 0.5.0
```
Comment (kyazdani42) [inline doc/nvim-tree-lua.txt:3]: good catch ^^'

File: lua/nvim-tree/events.lua [hunk 2]
```diff
@@ -0,0 +1,94 @@
+local M = {}
+
+local global_handlers = {}
+
+local Event = {
+    NodeRenamed = 'NodeRenamed',
+    FileCreated = 'FileCreated',
+    FileRemoved = 'FileRemoved',
+    FolderCreated = 'FolderCreated',
+    FolderRemoved = 'FolderRemoved',
```
Comment (kyazdani42) [inline lua/nvim-tree/events.lua:10]: could you indent this file with the `.editorconfig` settings please :) ? i use 2 spaces for indenting lua :)

**Review by kyazdani42 — CHANGES_REQUESTED** [review]
thanks for the PR, your lua code is quite pleasant to read. Just a small change and i'll merge this

File: lua/nvim-tree/events.lua (hunk 2 above)
Comment (williamboman) [reply lua/nvim-tree/events.lua]: :+1: There doesn't seem to be an `.editorconfig` so I added one, let me know if you want me to remove it

Comment (kyazdani42) [reply lua/nvim-tree/events.lua]: woups i thought i had one ^^' thanks for adding it

**Review by kyazdani42 — APPROVED** [review]
LGTM, thanks :)

**Pull request merged**